tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,512 @@
+			if (tid)
+				ts_catalog_update_tid_only(state->cagg_log_rel, tid, tup1);
+			else
+				ts_catalog_insert_only(state->cagg_log_rel, tup1);
+
+			heap_freetuple(tup1);
+		}
+
+		if (tup2)
+		{
+			ts_catalog_insert_only(state->cagg_log_rel, tup2);
+			heap_freetuple(tup2);
+		}
+
+		ts_catalog_restore_user(&sec_ctx);
+	}
+
+	return result;
+}
+
+/*
+ * Process invalidations in the hypertable invalidation log.
+ *
+ * Copy and delete all entries from the hypertable invalidation log.  For the
+ * continuous aggregate that is getting refreshed, we also match the
+ * invalidation against the refresh window and perform additional processing
+ * (cutting or deleting); work that we'd otherwise have to do later in the
+ * cagg invalidation log.
+ *
+ * Note that each entry gets one copy per continuous aggregate in the cagg
+ * invalidation log (unless it matched the refresh window). These copied
+ * entries are later used to track invalidations across refreshes on a
+ * per-cagg basis.
+ *
+ * After this function has run, there are no entries left in the hypertable
+ * invalidation log.
+ */
+static void
+move_invalidations_from_hyper_to_cagg_log(const CaggInvalidationState *state,
+										  const InternalTimeRange *refresh_window)
+{
+	int32 hyper_id = state->cagg.data.raw_hypertable_id;
+	List *cagg_ids = get_cagg_ids(hyper_id);
+	int32 last_cagg_hyper_id = llast_int(cagg_ids);
+	ScanIterator iterator;
+	ListCell *lc;
+
+	Assert(list_length(cagg_ids) > 0);
+
+	/* Switch to a per-tuple memory context since we could be processing a lot
+	 * of invalidations (basically an unbounded amount). */
+	MemoryContextReset(state->per_tuple_mctx);
+	hypertable_invalidation_scan_init(&iterator, hyper_id, RowExclusiveLock);
+	iterator.ctx.snapshot = state->snapshot;
+
+	/*
+	 * Looping over all continuous aggregates in the outer loop ensures all
+	 * tuples for a specific continuous aggregate is inserted consecutively in
+	 * the cagg invalidation log. This creates better locality for scanning
+	 * the invalidations later.
+	 */
+	foreach (lc, cagg_ids)
+	{
+		int32 cagg_hyper_id = lfirst_int(lc);
+
+		/* Scan all invalidations */
+		ts_scanner_foreach(&iterator)
+		{
+			TupleInfo *ti;
+			Form_continuous_aggs_hypertable_invalidation_log form;
+			CatalogSecurityContext sec_ctx;
+			MemoryContext oldmctx;
+			bool should_free;
+			HeapTuple tuple;
+			bool should_insert = false;
+
+			oldmctx = MemoryContextSwitchTo(state->per_tuple_mctx);
+			ti = ts_scan_iterator_tuple_info(&iterator);
+			tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
+			form = (Form_continuous_aggs_hypertable_invalidation_log) GETSTRUCT(tuple);
+
+			/* If we're processing an invalidation for the continuous
+			 * aggregate that is getting refreshed, then we can cut or delete
+			 * the invalidation immediately, instead of doing it later in the
+			 * cagg invalidation log. */
+			if (cagg_hyper_id == state->cagg.data.mat_hypertable_id)
```

Comment:
Is the scenario below a risk?

You get an invalidation range here and have a refresh window as follows:

```
      1           9
      +-----------+                   refresh window
           ++++++++++++               invalidation range
           5          20
```

You cut the invalidation range here and write it to the log

```
      1           9
      +-----------+                   refresh window
           ++++++++                   invalidation range
           5      9
```

Later, you move to processing it, but then the refresh window has moved:

```
      6           15
      +-----------+                   refresh window
    ++++++++                          invalidation range
    5      9
```

At this point, you cut the window again, but you've lost the invalidation in the range [9,20].

Reply:
This scenario doesn't exist because this is not how things are cut. Your example is wrong; invalidations are cleared inside the refresh window, not outside it like in your example. In other words, the part of an invalidation that remains is the part outside the refresh window. This is how things would be cut in your example:


```
      1           9
      +-----------+                   refresh window
           ++++++++++++               invalidation range
           5          20
```

You cut the invalidation range here and write it to the log

```
      1           9
      +-----------+                   refresh window
                          +++++               invalidation range
                       20
```

Reply:
Ok, but don't we have the opposite case then?

```
      5           9
      +-----------+        refresh window
++++++++++++               invalidation range
1          7
```

Which is then cut to:

```
      5           9
      +-----------+        refresh window
+++++++               invalidation range
1     5
```

And later:

```
          9           13
          +-----------+    refresh window
+++++++               invalidation range
1     5
```

Reply:
Nothing is lost in this case. The invalidation remains, even if you aren't refreshing a region that covers the invalidation. If you later refresh a region covering 1-5, that invalidation will be used to optimize that refresh (or at least that is the intention once such optimizations are implemented).